aquadoggo/src/materializer/tasks/schema.rs

```diff
@@ -1,9 +1,250 @@
 // SPDX-License-Identifier: AGPL-3.0-or-later
 
+use log::debug;
+use p2panda_rs::document::DocumentViewId;
+use p2panda_rs::operation::OperationValue;
+use p2panda_rs::schema::SchemaId;
+
 use crate::context::Context;
-use crate::materializer::worker::TaskResult;
+use crate::db::traits::{DocumentStore, SchemaStore};
+use crate::materializer::worker::{TaskError, TaskResult};
 use crate::materializer::TaskInput;
 
-pub async fn schema_task(_context: Context, _input: TaskInput) -> TaskResult<TaskInput> {
-    todo!()
+/// A schema task assembles and stores schemas from their views.
+///
+/// Schema tasks are dispatched whenever a schema definition or schema field definition document
+/// has all its immediate dependencies available in the store. It collects all required views for
```

Comment:
Ah right, that was not very clear of me!

You're indeed checking if all schema dependencies are met in the `dependency` task before dispatching `schema`.

I got confused because the flow in Whimsical is slightly different, where `dependency` is not doing the checks but `schema`. The reasoning to do the check inside of `schema` was a) what if schema gets dispatched somewhere else one day? Then you might need to write that check again (think about it as an API function you can call from anywhere) b) dispatching tasks (should be) inexpensive.

I'm fine with making this a ticket only for now where we revisit if it makes sense to do all checks _before_ or _after_ dispatching. It's an optimization and maybe needs more backup through benchmarks or reading some quening theory (it's only intuition from my side so far).

Reply:
> Ah, I see! Although, the schema task as it is implemented here will behave like expected and [fail if dependencies are not ready](https://github.com/p2panda/aquadoggo/pull/166/files#diff-796c5ead25350eadd5d7c3c571621befd33c2689411cf4d75ee329cf935c3b31R69-R71). There is no test for that though.

Ah, okay, that would be another good argument to move the checks in there there, as we might want to call `schema` from somewhere else as well some day (one could see it as a public function).

Reply:
Which checks do you mean? The schema task does "check", i.e. fail as expected, whether dependencies are met?

Reply:
Ah lol, I thought you've meant failing as in panicking, all good :+1:

Reply:
.. then we could also remove the checks in the `dependency` task and save one database query

Reply:
I'm not following this mega closely, but thought it was worth adding that when we first implemented reduce and dependency tasks there was no schema task to follow up, so maybe that is the cause of duplication popping up now.